pyccel/ast/numpyext.py

```diff
@@ -890,9 +894,10 @@ def __init__(self, shape, fill_value, dtype=None, order='C'):
         # If there is no dtype, extract it from fill_value
         # TODO: must get dtype from an annotated node
         if not dtype:
-            dtype = fill_value.dtype
-
+            dtype = DtypePrecisionToCastFunction[fill_value.dtype.name][fill_value.precision]
```

Comment:
Does `process_dtype` need to be called if the type is deduced from the array elements?

Reply:
Actually not.

Reply:
In that case I would rather we set the value in an if/else block. It is unnecessarily costly to use the desired dtype and precision to find the cast function which will ensure that `process_dtype` returns the desired dtype and precision

Reply:
When actually coding that if else statement, i've found myself repeating a lot of what's done in `process_dtype` especially since  a different treatment of `dtype` is done depending on the precision (python types and numpy types) , so maybe dtype and precision calculation should still pass by `process_type` ?

Reply:
Maybe you misunderstood me then. I was thinking of an if/else like:
```
if not dtype:
    dtype = fill_value.dtype
    precision = fill_value.precision
else:
    dtype, precision = process_type(dtype)
```

Reply:
Oh yeah that works ! assuming that the precision is valid.
```python
        if not dtype:
            dtype = fill_value.dtype
            precision = fill_value.precision
            if precision == -1:
                precision = default_precision[str(dtype)]
        else:
            dtype, precision = process_dtype(dtype)
```
Something like that maybe ?

Reply:
Or maybe:
```python
if not dtype:
    dtype = fill_value.dtype
    precision = get_final_precision(fill_value.precision)
else:
    dtype, precision = process_type(dtype)
```

Reply:
Yep !

Reply:
Fixed in [acd42c3](https://github.com/pyccel/pyccel/pull/1166/commits/acd42c3d3e928a51a9188f6b868e3068f71b83ca).